src/engine/DataObjects.hpp

```diff
@@ -0,0 +1,86 @@
+#pragma once
+
+namespace admirals {
+
+class Vector2 {
+protected:
+    float m_0;
+    float m_1;
```

Comment:
Because Vector3 inherits from Vector2, and Vector3 from Vector4, this makes each class inherit their parent class members. If had used a float array (etc float[2] for Vector2), then Vector3 would inherit a float[2] array that could not fit the data needed for a Vector3, if the Vector3 then declared its own member, it would then have both arrays and operations on the data could modify either of the two arrays depending on where the function is defined. 

An alternate solution would be to allocate memory on the heap and use a pointer inside the vector structs, in this case, the size could be determined dynamically.